File: src/main/java/com/cxxcxx/zinecraft/api/skill/SkillCatalog.java
package com.cxxcxx.zinecraft.api.skill;

import com.cxxcxx.zinecraft.api.item.ItemCatalog;
import com.cxxcxx.zinecraft.api.localization.TranslationCatalog;
import net.neoforged.neoforge.registries.DeferredItem;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public final class SkillCatalog {
  private final ItemCatalog items;
  private final TranslationCatalog translations;
  private final List<SkillEntry> mutableEntries = new ArrayList<>();

  public final List<SkillEntry> entries = Collections.unmodifiableList(mutableEntries);

  public SkillCatalog(ItemCatalog items, TranslationCatalog translations) {
    this.items = Objects.requireNonNull(items, "items");
    this.translations = Objects.requireNonNull(translations, "translations");
  }

  /**
   * 注册技能定义及其对应技能物品，并集中生成技能说明与 Ponder 文本。
   *
   * @param path            技能和技能物品共用的注册路径
   * @param zhCn            技能中文名
   * @param enUs            技能英文名
   * @param operatorZhCn    干员中文名
   * @param operatorEnUs    干员英文名
   * @param profession      技能所属职业
   * @param recoveryZhCn    中文技力回复类型
   * @param recoveryEnUs    英文技力回复类型
   * @param triggerZhCn     中文触发方式
   * @param triggerEnUs     英文触发方式
   * @param initialSp       初始技力，不能为负数
   * @param spCost          技力消耗，不能为负数
   * @param durationSeconds 持续秒数；瞬时技能传入 {@code null}
   * @param descriptionZhCn 中文技能描述
   * @param descriptionEnUs 英文技能描述
   * @param theme           Ponder 演示使用的视觉主题
   * @return 同时包含技能定义和注册物品的封装条目
   */
  public SkillEntry register(
      String path,
      String zhCn,
      String enUs,
      String operatorZhCn,
      String operatorEnUs,
      SkillProfession profession,
      String recoveryZhCn,
      String recoveryEnUs,
      String triggerZhCn,
      String triggerEnUs,
      int initialSp,
      int spCost,
      @Nullable Integer durationSeconds,
      String descriptionZhCn,
      String descriptionEnUs,
      SkillDemoTheme theme
  ) {
    if (initialSp < 0) throw new IllegalArgumentException("初始技力不能为负数");
    if (spCost < 0) throw new IllegalArgumentException("技力消耗不能为负数");
    if (durationSeconds != null && durationSeconds <= 0) {
      throw new IllegalArgumentException("技能持续时间必须大于 0");
    }
    if (entries.stream().anyMatch(entry -> entry.definition().getPath().equals(path))) {
      throw new IllegalArgumentException("技能 ID 重复: " + path);
    }

    SkillDefinition definition = new SkillDefinition(
        path,
        zhCn,
        enUs,
        operatorZhCn,
        operatorEnUs,
        profession,
        recoveryZhCn,
        recoveryEnUs,
        triggerZhCn,
        triggerEnUs,
        initialSp,
        spCost,
        durationSeconds,
        descriptionZhCn,
        descriptionEnUs,
        theme
    );
    DeferredItem<SkillItem> item = items.builder(path, zhCn, () -> new SkillItem(definition))
        .enUs(enUs)
        .hideCreativeTab()
        .build();
    registerTranslations(definition);

    SkillEntry entry = new SkillEntry(definition, item);
    mutableEntries.add(entry);
    return entry;
  }

  private void registerTranslations(SkillDefinition skill) {
    String tooltipKey = "item.zinecraft." + skill.getPath() + ".tooltip";
    translations.add(
        tooltipKey + ".operator",
        "干员：" + skill.getOperatorZhCn() + " · " + skill.getProfession().getZhCn(),
        "Operator: " + skill.getOperatorEnUs() + " · " + skill.getProfession().getEnUs()
    );
    translations.add(
        tooltipKey + ".activation",
        skill.getRecoveryZhCn() + " · " + skill.getTriggerZhCn(),
        skill.getRecoveryEnUs() + " · " + skill.getTriggerEnUs()
    );

    String durationZhCn = skill.getDurationSeconds() == null ? "" : " · 持续 " + skill.getDurationSeconds() + "秒";
    String durationEnUs = skill.getDurationSeconds() == null ? "" : " · Duration " + skill.getDurationSeconds() + "s";
    String statsZhCn = "初始 " + skill.getInitialSp() + " · 消耗 " + skill.getSpCost() + durationZhCn;
    String statsEnUs = "Initial " + skill.getInitialSp() + " · Cost " + skill.getSpCost() + durationEnUs;
    translations.add(tooltipKey + ".stats", statsZhCn, statsEnUs);
    translations.add(tooltipKey + ".description", skill.getDescriptionZhCn(), skill.getDescriptionEnUs());

    String ponderKey = "zinecraft.ponder.skill_demo_" + skill.getPath();
    translations.add(
        ponderKey + ".header",
        skill.getOperatorZhCn() + "：" + skill.getZhCn(),
        skill.getOperatorEnUs() + ": " + skill.getEnUs()
    );
    translations.add(
        ponderKey + ".text_1",
        skill.getOperatorZhCn() + "的" + skill.getProfession().getZhCn() + "技能",
        "A " + skill.getProfession().getEnUs() + " skill used by " + skill.getOperatorEnUs()
    );
    translations.add(
        ponderKey + ".text_2",
        skill.getRecoveryZhCn() + " · " + skill.getTriggerZhCn(),
        skill.getRecoveryEnUs() + " · " + skill.getTriggerEnUs()
    );
    translations.add(ponderKey + ".text_3", statsZhCn, statsEnUs);
    translations.add(ponderKey + ".text_4", skill.getDescriptionZhCn(), skill.getDescriptionEnUs());
    translations.add(
        ponderKey + ".text_5",
        "演示为 Minecraft 机制化表达，技能资料取自 PRTS。",
        "This is a Minecraft interpretation based on PRTS skill data."
    );
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/api/skill/SkillCatalog.kt
package com.cxxcxx.zinecraft.api.skill

import com.cxxcxx.zinecraft.api.item.ItemCatalog
import com.cxxcxx.zinecraft.api.item.ItemEntry
import com.cxxcxx.zinecraft.api.localization.TranslationCatalog
import net.minecraft.ChatFormatting
import net.minecraft.network.chat.Component
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Rarity
import net.minecraft.world.item.TooltipFlag
import net.minecraft.world.level.ItemLike

/** 将技能资料注册为物品，并自动生成名称和多行说明翻译。 */
class SkillCatalog(
  private val items: ItemCatalog,
  private val translations: TranslationCatalog
) {
  val entries = mutableListOf<SkillEntry>()

  fun register(
    path: String,
    zhCn: String,
    enUs: String,
    operatorZhCn: String,
    operatorEnUs: String,
    profession: SkillProfession,
    recoveryZhCn: String,
    recoveryEnUs: String,
    triggerZhCn: String,
    triggerEnUs: String,
    initialSp: Int,
    spCost: Int,
    durationSeconds: Int?,
    descriptionZhCn: String,
    descriptionEnUs: String,
    theme: SkillDemoTheme
  ): SkillEntry {
    require(initialSp >= 0) { "初始技力不能为负数" }
    require(spCost >= 0) { "技力消耗不能为负数" }
    require(durationSeconds == null || durationSeconds > 0) { "技能持续时间必须大于 0" }
    require(entries.none { it.definition.path == path }) { "技能 ID 重复: $path" }

    val definition = SkillDefinition(
      path, zhCn, enUs, operatorZhCn, operatorEnUs, profession,
      recoveryZhCn, recoveryEnUs, triggerZhCn, triggerEnUs,
      initialSp, spCost, durationSeconds, descriptionZhCn, descriptionEnUs, theme
    )
    val itemEntry = items.register(
      path,
      zhCn,
      enUs
    ) { SkillItem(definition) }
    registerTranslations(definition)
    return SkillEntry(definition, itemEntry).also(entries::add)
  }

  private fun registerTranslations(skill: SkillDefinition) {
    val prefix = "item.zinecraft.${skill.path}.tooltip"
    translations.add(
      "$prefix.operator",
      "干员：${skill.operatorZhCn} · ${skill.profession.zhCn}",
      "Operator: ${skill.operatorEnUs} · ${skill.profession.enUs}"
    )
    translations.add(
      "$prefix.activation",
      "${skill.recoveryZhCn} · ${skill.triggerZhCn}",
      "${skill.recoveryEnUs} · ${skill.triggerEnUs}"
    )
    val durationZh = skill.durationSeconds?.let { " · 持续 ${it}秒" }.orEmpty()
    val durationEn = skill.durationSeconds?.let { " · Duration ${it}s" }.orEmpty()
    translations.add(
      "$prefix.stats",
      "初始 ${skill.initialSp} · 消耗 ${skill.spCost}$durationZh",
      "Initial ${skill.initialSp} · Cost ${skill.spCost}$durationEn"
    )
    translations.add("$prefix.description", skill.descriptionZhCn, skill.descriptionEnUs)
    val ponderPrefix = "zinecraft.ponder.skill_demo_${skill.path}"
    translations.add(
      "$ponderPrefix.header",
      "${skill.operatorZhCn}：${skill.zhCn}",
      "${skill.operatorEnUs}: ${skill.enUs}"
    )
    translations.add(
      "$ponderPrefix.text_1",
      "${skill.operatorZhCn}的${skill.profession.zhCn}技能",
      "A ${skill.profession.enUs} skill used by ${skill.operatorEnUs}"
    )
    translations.add(
      "$ponderPrefix.text_2",
      "${skill.recoveryZhCn} · ${skill.triggerZhCn}",
      "${skill.recoveryEnUs} · ${skill.triggerEnUs}"
    )
    translations.add(
      "$ponderPrefix.text_3",
      "初始 ${skill.initialSp} · 消耗 ${skill.spCost}$durationZh",
      "Initial ${skill.initialSp} · Cost ${skill.spCost}$durationEn"
    )
    translations.add("$ponderPrefix.text_4", skill.descriptionZhCn, skill.descriptionEnUs)
    translations.add(
      "$ponderPrefix.text_5",
      "演示为 Minecraft 机制化表达，技能资料取自 PRTS。",
      "This is a Minecraft interpretation based on PRTS skill data."
    )
  }
}

class SkillItem internal constructor(
  val skill: SkillDefinition,
  properties: Properties = Properties().stacksTo(1).rarity(Rarity.RARE)
) : Item(properties) {
  override fun appendHoverText(
    stack: ItemStack,
    context: TooltipContext,
    tooltipComponents: MutableList<Component>,
    tooltipFlag: TooltipFlag
  ) {
    super.appendHoverText(stack, context, tooltipComponents, tooltipFlag)
    val prefix = descriptionId + ".tooltip"
    tooltipComponents += Component.translatable("$prefix.operator").withStyle(ChatFormatting.GOLD)
    tooltipComponents += Component.translatable("$prefix.activation").withStyle(ChatFormatting.AQUA)
    tooltipComponents += Component.translatable("$prefix.stats").withStyle(ChatFormatting.YELLOW)
    tooltipComponents += Component.translatable("$prefix.description").withStyle(ChatFormatting.GRAY)
  }
}

data class SkillEntry(
  val definition: SkillDefinition,
  val itemEntry: ItemEntry<SkillItem>
) : ItemLike {
  val item: SkillItem get() = itemEntry.item
  override fun asItem(): Item = item
}

data class SkillDefinition(
  val path: String,
  val zhCn: String,
  val enUs: String,
  val operatorZhCn: String,
  val operatorEnUs: String,
  val profession: SkillProfession,
  val recoveryZhCn: String,
  val recoveryEnUs: String,
  val triggerZhCn: String,
  val triggerEnUs: String,
  val initialSp: Int,
  val spCost: Int,
  val durationSeconds: Int?,
  val descriptionZhCn: String,
  val descriptionEnUs: String,
  val theme: SkillDemoTheme
)

enum class SkillProfession(val zhCn: String, val enUs: String) {
  VANGUARD("先锋", "Vanguard"),
  GUARD("近卫", "Guard"),
  SNIPER("狙击", "Sniper"),
  CASTER("术师", "Caster"),
  DEFENDER("重装", "Defender"),
  MEDIC("医疗", "Medic"),
  SUPPORTER("辅助", "Supporter"),
  SPECIALIST("特种", "Specialist")
}

enum class SkillDemoTheme {
  COST_RECOVERY,
  AREA_SLASH,
  RAPID_FIRE,
  VOLCANIC_BURST,
  HEAL_AND_SLOW,
  SANCTUARY,
  SLOWING_FIELD,
  DEPLOYMENT_STUN
}
